### ISSUE TITLE
http-client 0.4.9 ignores Connection: keep-alive

### ISSUE THREAD

**Issue by edom** [snoyberg/http-client#119]
I wrote a program to HTTP GET this URL (using defaultManagerSettings):

http://api.statdns.com/google.com/a

but after I got the response (using withResponse), Wireshark shows that the local machine initiates TCP FIN (implying that http-client closes the connection). Wireshark also shows that there is a Connection: keep-alive header in the response.

Setting the manager's timeout to Nothing also doesn't enable keepalives.

I thought http-client would handle keepalives? The documentation seems to imply that a Manager handles keepalives.

**Comment by snoyberg** [comment]
What code are you using to test this? I just checked against a patched version of http-client that prints a message when closing a socket, and used the following program:

```haskell
import Network.HTTP.Client

main :: IO ()
main = do
    putStrLn "parsing"
    req <- parseUrl "http://api.statdns.com/google.com/a"
    lbs <- withManager defaultManagerSettings $ \man -> do
        putStrLn "have a manager"
        withResponse req man $ \res -> do
            putStrLn "have a response"
            brConsume $ responseBody res
        putStrLn "second request"
        withResponse req man $ \res -> do
            putStrLn "have a second response"
            brConsume $ responseBody res
        putStrLn "finished response"
    putStrLn "finished manager"
```

The output is the following, which demonstrates that (1) the same connection is used for both requests and (2) the connection is only closed with the manager shuts down:

```
parsing               
have a manager        
have a response       
second request        
have a second response
finished response     
closing socket        
finished manager      
```

**Comment by edom** [comment]
I ran your code on my machine and the bug disappears, but the bug reappears if I replace all brConsume with brRead. Here is exactly the same code but with all brConsume replaced by brRead:

```haskell
import Network.HTTP.Client

main :: IO ()
main = do
    putStrLn "parsing"
    req <- parseUrl "http://api.statdns.com/google.com/a"
    lbs <- withManager defaultManagerSettings $ \man -> do
        putStrLn "have a manager"
        withResponse req man $ \res -> do
            putStrLn "have a response"
            brRead $ responseBody res
        putStrLn "second request"
        withResponse req man $ \res -> do
            putStrLn "have a second response"
            brRead $ responseBody res
        putStrLn "finished response"
    putStrLn "finished manager"
```

I'm using brRead as suggested by the documentation:

> You will need to use functions such as brRead to consume the response body.

and:

> If you are anticipating a large response body, you are encouraged to use withResponse and brRead instead.

**Comment by snoyberg** [comment]
I see what's happening here: by not calling `brRead` a second time, http-client isn't certain that the body has been fully consumed. Generally your usage of the library is incorrect, though in the case where the body fits into a single chunk, it will work. We can still probably improve http-client to more robustly detect when the entire body has been consumed.

tl;dr: You need to keep calling `brRead` until you get an empty `ByteString`.

**Comment by edom** [comment]
I see.

Assuming withResponse uses responseClose, will the following fix the problem?

* responseClose shall discard all data (brRead until empty).

What do you think?

**Comment by snoyberg** [comment]
It's a pessimization: the response body may be very large, which is why we don't reuse the connection in those cases.

**Comment by edom** [comment]
I see; it's a feature, not a bug: if the unread part is big, reconnecting is much quicker.

I see that you have added the documentation. Please feel free to close this.

Thanks!

**Comment by snoyberg** [comment]
No problem :)
